### PR TITLE
Polymer tool

### PR DESCRIPTION
**Pull request by chrisjonesBSU**
This is still a work in progress, but putting it out here for input, feedback, and help.

The basic framework is to break down a system of chain-like molecules into a hierarchical system of objects. An input would be a `.gsd` file. This relies quite heavily on the Freud cluster model (which we have a function for in `gsd_utils.py`)

The different classes:
```
System()
Molecule()
Segment()
Monomer()
```

The idea is that each one would sort of feed into the next.  For example

`System.molecules` would generate/return a list of `Molecule()` objects.
`Molecule.segments` or `Molecule.monomers` would generate/return a list of `Segment()` or `Monomer()` objects that belong to that specific molecule instance.

`Segment.monomers()` would generate a list of `Monomers()` that belong to that specific segment instance.

The `Segment` class is sort of optional, you could go straight from `Molecules()` to `Monomers()`

By extension both the `Segment()` and `Monomer()` objects would be available at the system level as well.

The goal is to easily track different features of individual molecules, segments and monomers and calculate system properties.

**Examples include:**
- Average end-to-end distance
- Average radius of gyration
- Distributions of both of the above
- Distribution of bond lengths between 2 monomers (or segments)
- Distribution of bond angles between 3 monomers (or segments)
- Mean square displacements of monomers or segments

I also think it could be helpful in creating coarse grain representations of the system using the monomer and/or segment classes which could be use in conjunction with the `MSIBI` workflow and `grits`.

**To Do:**
-------------------------------------
- [x] Use generators where possible rather than lists
- [x] Work on some of the system and molecule level property functions
- [x] The segment class still needs some logic, maybe an input of monomers per segment?
- [ ] Add unit tests
- [x] Add doc strings
- [x] CG compound generator, or interface with Grits
- [x] Add a `Component` class that can track the indices of monomer components like a branch or ring

### CODE REVIEW

**Comment by jennyfothergill** [comment]
First, this PR is an AWESOME idea, and I am excited to see what we can do with it! In working on grits I find that smarts string matching as the only method to generate a CG mapping leaves **a lot** to be desired, so having another method to accomplish this task (breaking a big thing into smaller things) is not overlapping--it instead gives us freedom to choose which method works best. Not to mention that this PR is adding a lot of new polymer specific functionality. :) I definitely want to make [grits mapping](https://github.com/cmelab/grits/issues/12) compatible with this code.

A suggestion I have about your outlined steps in the PR description: although generators may be more memory efficient, they may not necessarily be the best way to store iterable data. Whatever choice you make should be based on profiling the most common use cases. (e.g. do we commonly get the index of a particle? add and remove particles? iterate through the list?) Depending on how it is used, a generator may be slower and less convenient than a list. TLDR: wait until later to decide if you should replace lists with generators.

**Comment by chrisjonesBSU** [comment]
> First, this PR is an AWESOME idea, and I am excited to see what we can do with it! In working on grits I find that smarts string matching as the only method to generate a CG mapping leaves **a lot** to be desired, so having another method to accomplish this task (breaking a big thing into smaller things) is not overlapping--it instead gives us freedom to choose which method works best. Not to mention that this PR is adding a lot of new polymer specific functionality. :) I definitely want to make [grits mapping](https://github.com/cmelab/grits/issues/12) compatible with this code.
> 
> A suggestion I have about your outlined steps in the PR description: although generators may be more memory efficient, they may not necessarily be the best way to store iterable data. Whatever choice you make should be based on profiling the most common use cases. (e.g. do we commonly get the index of a particle? add and remove particles? iterate through the list?) Depending on how it is used, a generator may be slower and less convenient than a list. TLDR: wait until later to decide if you should replace lists with generators.

Good point, once more progress is made on this we can do some bench marking and test out different use cases : )

**Comment by chrisjonesBSU** [comment]
Per conversation with @jennyfothergill on Slack, we will probably create a general structure class that the others can inherit from.

```
class Structure():
    def __init__(self, system, molecule_id, atom_indices):
        self.system = system
        if molecule_id:
            self.atom_indices = np.where(self.system.clusters == molecule_id)[0]
        else:
            self.atom_indices = atom_indices
        self.n_atoms = len(self.atom_indices)

    @property
    def atom_positions(self):
        return self.system.snap.particles.position[self.atom_indices]

    @property
    def center_of_mass(self):
        freud_box = freud.Box(
                Lx = self.system.box[0],
                Ly = self.system.box[1],
                Lz = self.system.box[2]
                )
        return freud_box.center_of_mass(self.atom_positions)
```

**Comment by chrisjonesBSU** [comment]
I'm playing around with adding a `Component` class, so we can create structure groupings at a level smaller than a single monomer.  Here is an example with PEEK, where I made 3 components. 2 of them are a benzene ring with an oxygen, the other is a benzene ring with a carbon and oxygen.  

Maybe this can help with https://github.com/cmelab/grits/issues/16 ?

What's required is an index mapping scheme, that might be able to automatically be generated with what Grits already does?

For example, for PEEK I did this:

```
comp_mapping = {
    "ring_1": [0, 1, 2, 3, 4, 20, 21],
    "ring_2": [5, 6, 7, 8, 9, 18, 19],
    "ring_3": [10, 11, 12, 13, 14, 15, 16, 17]
}

```

Now, each component exists as its own object and we can use track their positions, add bonds between them, etc..

component_CG_example

**Comment by chrisjonesBSU** [comment]
> What's required is an index mapping scheme, that might be able to automatically be generated with what Grits already does?

Looking through GRiTS again, I see that the SMARTS matching does return these sub structure specific indices, but it still requires using an mBuild compound rather than staying purely in the realm of GSD files.

**Comment by chrisjonesBSU** [comment]
Here is an example of how this can be used to coarse-grain an atomistic gsd, and create a new GSD file with the CG'd representation.

This uses 1 monomer = 1 bead
```
gsdfile = "100-20mers.gsd"
peek_system = polymer.System(gsd_file=gsdfile, gsd_frame=0, atoms_per_monomer=22)

sequence = "PM"
for molecule in peek_system.molecules:
    molecule.assign_types(sequence=sequence)

snap = cmeutils.writers.write_snapshot(beads=[i for i in peek_system.monomers()])
cg_monomers_gsd = gsd.hoomd.open(name="cg_monomers_peek.gsd", mode="wb")
cg_monomers.append(snap)
```

If you wanted to CG at a level smaller than 1 monomer, you can use the `Components` class

```
gsdfile = "100-20mers.gsd"
peek_system = polymer.System(gsd_file=gsdfile, gsd_frame=0, atoms_per_monomer=22)
comp_mapping = {
    "A": [[0, 1, 2, 3, 4, 20, 21], [5, 6, 7, 8, 9, 18, 19]],
    "B": [10, 11, 12, 13, 14, 15, 16, 17]
}

for monomer in peek_system.monomers():
    monomer.generate_components(index_mapping = comp_mapping)

snap = cmeutils.writers.write_snapshot(beads=[i for i in peek_system.components()])
cg_components_gsd = gsd.hoomd.open(name="cg_rings_peek.gsd", mode="wb")
cg_components_gsd.append(snap)
```

File: cmeutils/gsd_utils.py [hunk 1]
```diff
@@ -32,6 +32,15 @@ def get_type_position(typename, gsd_file=None, snap=None, gsd_frame=-1):
...
     return typepos
 
+def unwrap_coordinates(gsd_file=None, snap=None, gsd_frame=-1):
+    snap = _validate_inputs(gsd_file, snap, gsd_frame)
+    box = snap.configuration.box
+    if box[3:].any():
+        raise ValueError("The unwrap_coordinates function only works "
+                "with orthogonal boxes."
+                )
+    pos = snap.particles.position + (snap.particles.image * box[:3])
```
Comment (jennyfothergill) [inline cmeutils/gsd_utils.py:42]: Not all gsd snapshots will have the `image` attribute.  Only those with the `dynamic=['momentum']` kwarg passed to `hoomd.dump.gsd()`. We should add some error handling for those without this. (I can't remember the value of `snap.particles.image` if momentum is not provided...)

**Comment by chrisjonesBSU** [comment]
Putting this PR on pause for now; we're seeing how/if some of this funcitonality can live in [Grits](https://github.com/cmelab/grits) and this PR will be changed to contain only the analysis related stuff, like end-to-end distance, radius of gyration, etc.  Maybe it will rely only Grits to provide information about the structural components of the system at a level larger than individual atoms.

File: cmeutils/writers.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
+import gsd
+import gsd.hoomd
+import hoomd
```
Comment (jennyfothergill) [inline cmeutils/writers.py:3]: We probably should add hoomd to the environment yml

**Comment by chrisjonesBSU** [comment]
For now, I'm just going to move this to it's own repo (https://github.com/chrisjonesBSU/paek_cg).

The more I work on it, the more assumptions I make about the polymer structure, and since I'm working on this with PEEK and PEKK in mind it's not really designed to work with polymers that have side chains, functional groups, etc. Even a simple system like P3HT wouldn't work unless the corase-graining was at a level of 1 monomer = 1 bead or larger.  So, I'm not sure if it belongs in `cmeutils` if it's becoming more and more tailored to a specific system.

**Comment by chrisjonesBSU** [comment]
See comment above

**Pull request closed without merging**